lib/benchpark/cuda.py

```diff
@@ -0,0 +1,131 @@
+      1. **The experiment class**.  Classes contain ``directives``, which are
+         special functions, that add metadata (variants) to packages (see
+         ``directives.py``).
+
+      2. **Experiment instances**. Once instantiated, an experiment is
+         essentially a collection of files defining an experiment in a
+         Ramble workspace.
+    """
+
+    variant("cuda", default=False, description="Build and run with CUDA")
+
+
+
+    #
+    # These are default values for instance variables.
+    #
+
+    # This allows analysis tools to correctly interpret the class attributes.
+    variants: Dict[
+        "benchpark.spec.Spec",
+        Dict[str, benchpark.variant.Variant],
+    ]
+
+    def __init__(self, spec):
+        self.spec: "benchpark.spec.ConcreteExperimentSpec" = spec
+        super().__init__()
+
+    def compute_include_section(self):
+        # include the config directory
+        # TODO: does this need to change to interop with System class
+        return ["./configs"]
+
+    def compute_config_section(self):
+        # default configs for all experiments
+        return {
+            "deprecated": True,
+            "spack_flags": {"install": "--add --keep-stage", "concretize": "-U -f"},
+        }
+
+    def compute_modifiers_section(self):
+        # by default we use the allocation modifier and no others
+        return [{"name": "allocation"}]
+
+    def compute_applications_section(self):
+        # TODO: is there some reasonable default?
+        variables = {}
+        variables["n_gpus"] = num_procs  # TODO: num_procs will be defined in the child...
+
+        raise NotImplementedError(
+            "Each experiment must implement compute_applications_section"
+        )
+
+    def compute_spack_section(self):
+        # TODO: is there some reasonable default based on known variable names?
+        system_specs = {}
+        system_specs["cuda_version"] = "{default_cuda_version}"
+        system_specs["cuda_arch"] = "{cuda_arch}"
+        system_specs["blas"] = "cublas-cuda"  #TODO: can we define this in Lapack/Blas CudaExperiment?
+
+        package_specs = {}
+        package_specs["cuda"] = {
+            "pkg_spec": "cuda@{}+allow-unsupported-compilers".format(
+                system_specs["cuda_version"]
+            ),
+        "compiler": system_specs["compiler"],
+        }
+
+        # TODO: can we define this in Lapack/Blas CudaExperiment?
+        package_specs[system_specs["blas"]] = (
+            {}
+        )  # empty package_specs value implies external package
+
+        # TODO: can we define this in Hypre CudaExperiment?
+        package_specs["hypre"]["pkg_spec"] += "+cuda cuda_arch={}".format(
+            system_specs["cuda_arch"]
+        )
+        package_specs[app_name]["pkg_spec"] += "+cuda cuda_arch={}".format(
+            system_specs["cuda_arch"]
+        )
+
+        raise NotImplementedError(
+            "Each experiment must implement compute_spack_section"
+        )
```

Comment:
With an inheritance scheme, this would have to return something, e.g.

```
        return {
            "packages": {k: v for k, v in package_specs.items() if v},
            "environments": {app_name: {"packages": list(package_specs.keys())}},
        }
```

(copied from Amg2023) at which point, a subclass would want to do something like:

```
class FooExperiment(CudaExperiment):
  def compute_spack_section(self):
    config = super().compute_spack_section()
    # now add all the things specific to Foo
```

Using inheritance is somewhat inflexible, because everything you want to mix together would need to be part of a single inheritance chain (which is why I suggested composition in https://github.com/LLNL/benchpark/pull/356#discussion_r1783579262, for example).